lib/loader.js

```diff
@@ -0,0 +1,87 @@
+/*
+ * (C) Copyright 2016 o2r project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+var c = require('../config/config');
+const debug = require('debug')('loader');
+const Docker = require('dockerode');
+const steps = require('../lib/loader-steps');
+const randomstring = require('randomstring');
+
+/**
+ * Create Loader to handle given request and response
+ * @constructor
+ * @param {object} request - The load request
+ * @param {object} response - The response
+ */
+function Loader(req, res) {
+    this.req = req;
+    this.res = res;
+
+    this.load = (done) => {
+        debug('Handling public share load of %s for user %s', this.req.body.share_url, this.req.user.orcid);
+
+        let passon = {
+            id: randomstring.generate(c.id_length),
+            shareURL: encodeURI(this.req.body.share_url),
+            webdav_path: this.req.body.path,
+            user: this.req.user.orcid,
+            req: this.req,
+            res: this.res,
+            docker: new Docker() // setup Docker client with default options
+        };
+
+        debug('[%s] Docker client set up: %s', passon.id, JSON.stringify(passon.docker));
+
+        // TODO: decide what to do based on file structure
+        
+        return steps.publicShareLoad(passon)
+            .then(steps.unzip)
+            .then(steps.extractMetadata)
+            .then(steps.loadMetadata)
+            .then(steps.brokerMetadata)
+            .then(steps.save)
+            .then(steps.cleanup)
+            .then(this.respond)
+            .then((passon) => {
+                debug('[%s] completed load', passon.id);
+                done(passon.id);
+            })
+            .catch(err => {
+                debug('[%s] Rejection or unhandled failure during execute: \n\t%s',
+                    this.passon.id, JSON.stringify(err));
+                let status = 500;
+                if (err.status) {
+                    status = err.status;
+                }
+                let msg = 'Internal error';
+                if (err.msg) {
+                    msg = err.msg;
+                }
+                res.status(status).send(JSON.stringify({ error: msg }));
```

Comment:
Yes, otherwise the callback is never called.

You are correct, I probably just got lucky in the muncher so far, just updated it: https://github.com/nuest/o2r-muncher/commit/0e3d738aa020d1c03eee84b7029f64153f5a4d18